### ISSUE TITLE
Writing optimizers.

### ISSUE THREAD

**Issue by oysteijo** [oysteijo/simd_neuralnet#10]
It looks like the optimizer code actually makes a difference of how low your error rate can go. In the optimizers directory, we have to make several optimizers. SGD, RMSprop, Adam and some others. Let's see.

**Comment by oysteijo** [comment]
Now that we have the basic foundation of the optimizers, I think I have to read up on the theories.

**Comment by oysteijo** [comment]
OK, we got SGD and Adagrad, I also think RMSprop is within reach. Adam is a bit harder, but doable, I believe.

**Comment by oysteijo** [comment]
RMSprop done, but w/o Nesterov. There is an issue in Nesterov in SGD s well. I'll look at that in a day or two.

**Comment by oysteijo** [comment]
OK with 4f2beb1 I've committed an implementation of Adam. It seems to work, as it gives the same-ish numbers as Keras.

That completes this list:

- SGD + momentum (Nesterov semi-works)
- AdaGrad
- RMSprop (no Nesterov available yet)
- Adam

I think that should do it for now. I will close this case. However, we need a cleanup. We need a bugfix in Nesterov (I will create a separate issue). And then there are other optimizers out there:

- [x] - AdamW (That is Adam with weight decay and L2 regulatization)
- [ ] - Nadam (That is Adam with Nesterov momentum)
- [ ] - Adamax
- [ ] - Amsgrad
- [ ] - Other (Just read up on new theories)

**Comment by oysteijo** [comment]
AdamW, as based on Ilya Loshchilov & Frank Hutter, Decoupled Weight Decay Regularization, has been implemented.

I'm trying this out on the backgammon dataset, but it does not seem to improve much.

Update: It does improve. It rocks!

**Comment by oysteijo** [comment]
The Ilya Loshchilov & Frank Hutte article describes SGDW, which is basically Stochastic Gradient Descent with Weight Decay. Should be really simple to implement.
